### PR TITLE
Wasm hooks E2E test & change intermediary address prefix

### PR DESCRIPTION
**Pull request by nicolaslara**
## What is the purpose of the change

Adds an e2e test to wasm hooks

## Brief Changelog

Adds an e2e test to wasm hooks

## Testing and Verifying
att tests pass

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: tests/e2e/e2e_test.go [hunk 1]
```diff
@@ -228,6 +230,78 @@ func (s *IntegrationTestSuite) TestIBCTokenTransferRateLimiting() {
...
+	s.NoError(err)
+	nodeB, err := chainB.GetDefaultNode()
+	s.NoError(err)
+
+	// copy the contract from x/rate-limit/testdata/
+	wd, err := os.Getwd()
+	s.NoError(err)
+	// co up two levels
+	projectDir := filepath.Dir(filepath.Dir(wd))
+	fmt.Println(wd, projectDir)
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:250]:
```suggestion
	// go up two levels
	projectDir := filepath.Dir(filepath.Dir(wd))
```

File: tests/e2e/e2e_test.go [hunk 2]
```diff
@@ -228,6 +230,78 @@ func (s *IntegrationTestSuite) TestIBCTokenTransferRateLimiting() {
...
+
+	nodeA.StoreWasmCode("counter.wasm", initialization.ValidatorWalletName)
+	chainA.LatestCodeId = int(nodeA.QueryLatestWasmCodeID())
+	nodeA.InstantiateWasmContract(
+		strconv.Itoa(chainA.LatestCodeId),
+		`{"count": 0}`,
+		initialization.ValidatorWalletName)
+
+	// Using code_id 1 because this is the only contract right now. This may need to change if more contracts are added
+	contracts, err := nodeA.QueryContractsFromId(chainA.LatestCodeId)
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:262]: I think this comment can be removed now right?

File: tests/e2e/e2e_test.go [hunk 3]
```diff
@@ -228,6 +230,78 @@ func (s *IntegrationTestSuite) TestIBCTokenTransferRateLimiting() {
...
+		fmt.Sprintf(`{"wasm":{"contract":"%s","msg": {"increment": {}} }}`, contractAddr))
+
+	// check the balance of the contract
+	s.Eventually(func() bool {
+		balance, err := nodeA.QueryBalances(contractAddr)
+		s.Require().NoError(err)
+		if len(balance) == 0 {
+			return false
+		}
+		return balance[0].Amount.Int64() == 10
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:279]: Where does 10 come from? Can this be not hard coded or does it make the most sense to do it this way?

File: tests/e2e/e2e_test.go [hunk 4]
```diff
@@ -228,6 +230,78 @@ func (s *IntegrationTestSuite) TestIBCTokenTransferRateLimiting() {
...
+	fmt.Println("sender", senderBech32)
+
+	var response map[string]interface{}
+	s.Eventually(func() bool {
+		response, err = nodeA.QueryWasmSmart(contractAddr, fmt.Sprintf(`{"get_total_funds": {"addr": "%s"}}`, senderBech32))
+		totalFunds := response["total_funds"].([]interface{})[0]
+		amount := totalFunds.(map[string]interface{})["amount"].(string)
+		denom := totalFunds.(map[string]interface{})["denom"].(string)
+		// check if denom contains "uosmo"
+		return err == nil && amount == "10" && strings.Contains(denom, "ibc")
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:296]: same question about 10

File: x/ibc-hooks/README.md [hunk 5]
```diff
@@ -34,7 +34,7 @@ So we detail where we want to get each of these fields from:
 * Sender: We cannot trust the sender of an IBC packet, the counterparty chain has full ability to lie about it. 
 We cannot risk this sender being confused for a particular user or module address on Osmosis.
 So we replace the sender with an account to represent the sender prefixed by the channel and a wasm module prefix.
-This is done by setting the sender to `Bech32(Hash("ibc-memo-action" || channelID || sender))`, where the channelId is the channel id on the local chain. 
+This is done by setting the sender to `Bech32(Hash("ibc-wasm-hook-intermediaryg" || channelID || sender))`, where the channelId is the channel id on the local chain.
```
Comment (czarcas7ic) [inline x/ibc-hooks/README.md:37]: I think there is a typo in "wasm-hook-intermediary" with the g at the end

File: x/ibc-hooks/client/cli/query.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package cli
```
Comment (czarcas7ic) [inline x/ibc-hooks/client/cli/query.go:1]: can we use osmocli for this instead?

**Review by czarcas7ic — APPROVED** [review]

File: tests/e2e/e2e_test.go [hunk 7]
```diff
@@ -228,6 +230,78 @@ func (s *IntegrationTestSuite) TestIBCTokenTransferRateLimiting() {
...
+		initialization.ValidatorWalletName)
+
+	// Using code_id 1 because this is the only contract right now. This may need to change if more contracts are added
+	contracts, err := nodeA.QueryContractsFromId(chainA.LatestCodeId)
+	s.NoError(err)
+	s.Require().Len(contracts, 1, "Wrong number of contracts for the counter")
+	contractAddr := contracts[0]
+
+	validatorAddr := nodeB.GetWallet(initialization.ValidatorWalletName)
+	fmt.Println("validatorAddr", validatorAddr)
```
Comment (p0mvn) [inline tests/e2e/e2e_test.go:268]:
```suggestion
```

File: tests/e2e/e2e_test.go (hunk 3 above)
Comment (p0mvn) [reply tests/e2e/e2e_test.go:279]: I think this is from the ibs transfer call up above. +1 to have a variable that we can format into the ibc transfer command and use here

File: tests/e2e/e2e_test.go [hunk 8]
```diff
@@ -228,6 +230,78 @@ func (s *IntegrationTestSuite) TestIBCTokenTransferRateLimiting() {
...
+		}
+		return balance[0].Amount.Int64() == 10
+	},
+		1*time.Minute,
+		10*time.Millisecond,
+	)
+
+	// sender wasm addr
+	senderBech32, err := ibchookskeeper.DeriveIntermediateSender("channel-0", validatorAddr, "osmo")
+	fmt.Println("sender", senderBech32)
```
Comment (p0mvn) [inline tests/e2e/e2e_test.go:287]: nit: I'm not sure if prints are helpful for debugging purposes. If so, can we change them to be logs. Logs format lines and test case info, making it easier to debug and also don't bypass testing package's output buffering
```suggestion
	s.T().Log("sender", senderBech32)
```

File: tests/e2e/e2e_test.go [hunk 9]
```diff
@@ -228,6 +230,78 @@ func (s *IntegrationTestSuite) TestIBCTokenTransferRateLimiting() {
...
+		totalFunds := response["total_funds"].([]interface{})[0]
+		amount := totalFunds.(map[string]interface{})["amount"].(string)
+		denom := totalFunds.(map[string]interface{})["denom"].(string)
+		// check if denom contains "uosmo"
+		return err == nil && amount == "10" && strings.Contains(denom, "ibc")
+	},
+		15*time.Second,
+		10*time.Millisecond,
+	)
+	fmt.Println("response", response)
```
Comment (p0mvn) [inline tests/e2e/e2e_test.go:301]:
```suggestion
	s.T().Log("response", response)
```

**Review by p0mvn — APPROVED** [review]
LGTM

**Review by ValarDragon — APPROVED** [review]

File: tests/e2e/e2e_test.go (hunk 8 above)
Comment (nicolaslara) [reply tests/e2e/e2e_test.go:287]: prints were a leftover from when I was debugging. Can be removed (or logged)

File: x/ibc-hooks/client/cli/query.go (hunk 6 above)
Comment (nicolaslara) [reply x/ibc-hooks/client/cli/query.go:1]: I started looking into that and realized I didn't know how to use osmocli 😅, so in the interest of time, I just went with what I knew

**Pull request merged**